File: app/lib/core/constants/firestore_paths.dart
/// Single source of truth for every Firestore collection/document path.
///
/// Why this exists: a typo'd path string (`'shcools'`) doesn't throw --
/// Firestore just silently creates/reads a different, empty location.
/// Routing every access through these builders makes path typos a
/// compile-time problem instead of a "why is my data missing" bug report.
///
/// Mirrors functions/src/shared/firestore-paths.ts -- keep both in sync.
class FirestorePaths {
  FirestorePaths._();

  // ---- Platform-level (Owner only) ----
  static const String platformSchools = 'platform_schools';
  static String platformSchoolDoc(String schoolId) => '$platformSchools/$schoolId';

  static const String platformSubscriptions = 'platform_subscriptions';
  static String platformSubscriptionDoc(String schoolId) =>
      '$platformSubscriptions/$schoolId';

  static const String platformInvoices = 'platform_invoices';
  static String platformInvoiceDoc(String invoiceId) => '$platformInvoices/$invoiceId';

  static const String platformOwnerAuditLog = 'platform_owner_audit_log';

  static const String ownerProfiles = 'platform_owner_profiles';
  static String ownerProfileDoc(String uid) => '$ownerProfiles/$uid';

  // ---- Tenant root ----
  static String school(String schoolId) => 'schools/$schoolId';

  static String users(String schoolId) => '${school(schoolId)}/users';
  static String userDoc(String schoolId, String userId) => '${users(schoolId)}/$userId';

  static String students(String schoolId) => '${school(schoolId)}/students';
  static String studentDoc(String schoolId, String studentId) =>
      '${students(schoolId)}/$studentId';

  static String employees(String schoolId) => '${school(schoolId)}/employees';

  static String attendance(String schoolId) => '${school(schoolId)}/attendance';

  static String payments(String schoolId) => '${school(schoolId)}/payments';

  /// Claims of online payment awaiting a cashier's verification. Separate
  /// from `payments`, which only ever holds money the school has confirmed.
  static String paymentSubmissions(String schoolId) =>
      '${school(schoolId)}/paymentSubmissions';

  /// Single document: the school's e-wallet QR and account details.
  static String paymentSettingsDoc(String schoolId) =>
      '${school(schoolId)}/settings/payments';

  /// Single document: logo and school name for the app and printed IDs.
  static String brandingDoc(String schoolId) => '${school(schoolId)}/settings/branding';

  /// The school's published fee schedules, and the record of each
  /// occasion one was charged to a student.
  static String feeStructures(String schoolId) => '${school(schoolId)}/feeStructures';

  static String assessments(String schoolId) => '${school(schoolId)}/assessments';

  static String programs(String schoolId) => '${school(schoolId)}/programs';

  static String inventory(String schoolId) => '${school(schoolId)}/inventory';
  static String inventoryTransactions(String schoolId) =>
      '${school(schoolId)}/inventoryTransactions';

  static String announcements(String schoolId) => '${school(schoolId)}/announcements';

  static String meetings(String schoolId) => '${school(schoolId)}/meetings';

  static String approvals(String schoolId) => '${school(schoolId)}/approvals';

  static String expenses(String schoolId) => '${school(schoolId)}/expenses';

  static String teacherAssignments(String schoolId) => '${school(schoolId)}/teacherAssignments';

  /// The weekly timetable. Written only by saveScheduleBlock and
  /// deleteScheduleBlock -- clash detection is the whole feature, and a
  /// client that could write here directly could book two classes into
  /// one room.
  static String scheduleBlocks(String schoolId) => '${school(schoolId)}/scheduleBlocks';

  /// What a person asked the school to do with their own information,
  /// and what the school did about it. Append-and-answer: the request
  /// itself never changes, and nothing is ever deleted -- a school asked
  /// how it handles these has to be able to show the refusals too.
  static String dataRequests(String schoolId) => '${school(schoolId)}/dataRequests';

  static String courseworkItems(String schoolId) => '${school(schoolId)}/courseworkItems';

  /// What students hand in against a courseworkItem. A sibling
  /// collection rather than a subcollection: a teacher's "who has
  /// submitted" view and a student's "what have I handed in" view are
  /// both flat queries across many items, which a subcollection would
  /// turn into a collectionGroup query and a second index.
  static String courseworkSubmissions(String schoolId) =>
      '${school(schoolId)}/courseworkSubmissions';

  /// The correct answers. Never readable by students -- see
  /// firestore.rules. Keyed by courseworkId so there is exactly one key
  /// per item and no way to end up with two disagreeing.
  static String courseworkAnswerKeys(String schoolId) =>
      '${school(schoolId)}/courseworkAnswerKeys';

  /// Published emergency numbers (BFP, PNP, clinic). Readable by every
  /// role in the school -- a number nobody can reach is not a safety
  /// feature.
  static String emergencyContacts(String schoolId) =>
      '${school(schoolId)}/emergencyContacts';

  /// Emergency alerts raised by students.
  static String emergencyAlerts(String schoolId) =>
      '${school(schoolId)}/emergencyAlerts';

  static String grades(String schoolId) => '${school(schoolId)}/grades';

  static String checklistItems(String schoolId) => '${school(schoolId)}/checklistItems';

  static String dailyReports(String schoolId) => '${school(schoolId)}/dailyReports';

  static String guidanceRecords(String schoolId) => '${school(schoolId)}/guidanceRecords';

  static String summons(String schoolId) => '${school(schoolId)}/summons';

  /// Per-subject attendance: the class the teacher opened, and the marks
  /// taken in it. Two collections rather than marks nested in the
  /// session, because "was this child in Physics all term" is a query
  /// across sessions, and a subcollection cannot be asked that without a
  /// collection-group index the marks do not otherwise need.
  static String classSessions(String schoolId) => '${school(schoolId)}/classSessions';
  static String subjectAttendance(String schoolId) =>
      '${school(schoolId)}/subjectAttendance';

  /// Employee leave. A collection of its own rather than a row in the
  /// generic approvals queue: the timesheet has to ask "was this day
  /// covered", which is a date-range query, and an employee has to be
  /// able to read their own without being able to read the school's.
  static String leaveRequests(String schoolId) => '${school(schoolId)}/leaveRequests';

  /// Parent-teacher messaging. Messages are a subcollection of their
  /// conversation because they are only ever read through it: there is
  /// no "all messages in the school" screen, and there must not be.
  static String conversations(String schoolId) => '${school(schoolId)}/conversations';
  static String messages(String schoolId, String conversationId) =>
      '${conversations(schoolId)}/$conversationId/messages';

  static String auditLog(String schoolId) => '${school(schoolId)}/auditLog';

  static String documents(String schoolId) => '${school(schoolId)}/documents';

  /// The log of TORs and Form 137s handed over the counter. Separate
  /// from `documents`, which holds files the school stores; this holds
  /// the record of a physical release.
  static String documentReleases(String schoolId) =>
      '${school(schoolId)}/documentReleases';

  static String notificationItems(String schoolId, String userId) =>
      '${school(schoolId)}/notifications/$userId/items';

  /// Per-school sequential counters (studentNumber, receiptNumber, etc.),
  /// incremented atomically inside Cloud Function transactions.
  static String counters(String schoolId) => '${school(schoolId)}/counters';
  static String counterDoc(String schoolId, String counterName) =>
      '${counters(schoolId)}/$counterName';
}

File: functions/src/shared/firestore-paths.ts
/**
 * Single source of truth for Firestore paths on the Functions side.
 * Mirrors app/lib/core/constants/firestore_paths.dart -- keep both in sync
 * whenever a collection is added or renamed.
 */
export const FirestorePaths = {
  platformSchools: "platform_schools",
  platformSchoolDoc: (schoolId: string) => `platform_schools/${schoolId}`,

  platformSubscriptions: "platform_subscriptions",
  platformSubscriptionDoc: (schoolId: string) => `platform_subscriptions/${schoolId}`,

  platformInvoices: "platform_invoices",
  platformInvoiceDoc: (invoiceId: string) => `platform_invoices/${invoiceId}`,

  platformOwnerAuditLog: "platform_owner_audit_log",

  ownerProfiles: "platform_owner_profiles",
  ownerProfileDoc: (uid: string) => `platform_owner_profiles/${uid}`,

  school: (schoolId: string) => `schools/${schoolId}`,

  users: (schoolId: string) => `schools/${schoolId}/users`,
  userDoc: (schoolId: string, userId: string) => `schools/${schoolId}/users/${userId}`,

  students: (schoolId: string) => `schools/${schoolId}/students`,
  studentDoc: (schoolId: string, studentId: string) =>
    `schools/${schoolId}/students/${studentId}`,

  // Employee leave. Its own collection rather than a row in the generic
  // approvals queue: the timesheet asks "was this day covered", which is
  // a date-range query, and an employee reads their own without reading
  // the school's.
  leaveRequests: (schoolId: string) => `schools/${schoolId}/leaveRequests`,

  // Parent-teacher messaging. Messages are a subcollection because
  // they are only ever read through their conversation -- there is no
  // "all messages in the school" screen, and there must not be.
  conversations: (schoolId: string) => `schools/${schoolId}/conversations`,
  conversationDoc: (schoolId: string, conversationId: string) =>
    `schools/${schoolId}/conversations/${conversationId}`,
  messages: (schoolId: string, conversationId: string) =>
    `schools/${schoolId}/conversations/${conversationId}/messages`,

  auditLog: (schoolId: string) => `schools/${schoolId}/auditLog`,

  attendance: (schoolId: string) => `schools/${schoolId}/attendance`,
  attendanceDoc: (schoolId: string, recordId: string) => `schools/${schoolId}/attendance/${recordId}`,

  // Per-subject attendance. Two collections rather than one, because a
  // session is a thing that happened to a class (it started at 7:32 and
  // ran 48 minutes) and a mark is a thing that happened to a student.
  // Folding the marks into the session document would cap a section at
  // whatever fits in one megabyte and make "was this child in Physics
  // all term" a scan of every session in the school.
  classSessions: (schoolId: string) => `schools/${schoolId}/classSessions`,
  classSessionDoc: (schoolId: string, sessionId: string) =>
    `schools/${schoolId}/classSessions/${sessionId}`,
  subjectAttendance: (schoolId: string) => `schools/${schoolId}/subjectAttendance`,
  subjectAttendanceDoc: (schoolId: string, markId: string) =>
    `schools/${schoolId}/subjectAttendance/${markId}`,

  // The BIR-registered booklets a school issues official receipts from,
  // and one claim document per number used. The claim's id IS the
  // number, which is what makes "one receipt, issued once" a guarantee
  // rather than a query two concurrent cashiers could both pass.
  receiptBooklets: (schoolId: string) => `schools/${schoolId}/receiptBooklets`,
  receiptClaims: (schoolId: string, bookletId: string) =>
    `schools/${schoolId}/receiptBooklets/${bookletId}/claims`,
  receiptClaimDoc: (schoolId: string, bookletId: string, number: string) =>
    `schools/${schoolId}/receiptBooklets/${bookletId}/claims/${number}`,

  payments: (schoolId: string) => `schools/${schoolId}/payments`,
  paymentDoc: (schoolId: string, paymentId: string) => `schools/${schoolId}/payments/${paymentId}`,
  // Claims of online payment awaiting a cashier's verification. Kept apart
  // from `payments`, which only ever holds money the school has confirmed.
  paymentSubmissions: (schoolId: string) => `schools/${schoolId}/paymentSubmissions`,

  // The school's published fee schedules, and the record of each
  // occasion one was charged to a student. Two collections, because a
  // schedule is a template the school edits between years and an
  // assessment is a thing that happened to a family.
  feeStructures: (schoolId: string) => `schools/${schoolId}/feeStructures`,
  feeStructureDoc: (schoolId: string, id: string) => `schools/${schoolId}/feeStructures/${id}`,
  assessments: (schoolId: string) => `schools/${schoolId}/assessments`,
  assessmentDoc: (schoolId: string, id: string) => `schools/${schoolId}/assessments/${id}`,

  programs: (schoolId: string) => `schools/${schoolId}/programs`,
  programDoc: (schoolId: string, programId: string) => `schools/${schoolId}/programs/${programId}`,

  teacherAssignments: (schoolId: string) => `schools/${schoolId}/teacherAssignments`,

  scheduleBlocks: (schoolId: string) => `schools/${schoolId}/scheduleBlocks`,
  scheduleBlockDoc: (schoolId: string, blockId: string) =>
    `schools/${schoolId}/scheduleBlocks/${blockId}`,
  emergencyAlerts: (schoolId: string) => `schools/${schoolId}/emergencyAlerts`,

  courseworkSubmissions: (schoolId: string) => `schools/${schoolId}/courseworkSubmissions`,
  // The correct answers. Never readable by students -- see firestore.rules.
  courseworkAnswerKeys: (schoolId: string) => `schools/${schoolId}/courseworkAnswerKeys`,

  counterDoc: (schoolId: string, counterName: string) =>
    `schools/${schoolId}/counters/${counterName}`,

  // ---- Admissions ----
  // Enquiries, from the first phone call to the day they enrol.
  applicants: (schoolId: string) => `schools/${schoolId}/applicants`,
  applicantDoc: (schoolId: string, applicantId: string) =>
    `schools/${schoolId}/applicants/${applicantId}`,

  // ---- Year-end rollover ----
  // One document per school year, holding what was decided and by whom.
  schoolYears: (schoolId: string) => `schools/${schoolId}/schoolYears`,
  schoolYearDoc: (schoolId: string, schoolYear: string) =>
    `schools/${schoolId}/schoolYears/${schoolYear}`,

  // One per student per year. The id is deliberate: it is what makes a
  // rollover safe to re-run, since a student already moved cannot be
  // created again.
  promotions: (schoolId: string) => `schools/${schoolId}/promotions`,
  promotionDoc: (schoolId: string, schoolYear: string, studentId: string) =>
    `schools/${schoolId}/promotions/${schoolYear}_${studentId}`,
};
